sale_start_end_dates/models/sale_order.py

```diff
@@ -0,0 +1,134 @@
+    default_end_date = fields.Date(string='Default End Date')
+
+    @api.constrains('default_start_date', 'default_end_date')
+    def _check_default_start_end_dates(self):
+        if (self.default_start_date and self.default_end_date and
+                self.default_start_date > self.default_end_date):
+            raise ValidationError(_(
+                "Default Start Date should be before or be the "
+                "same as Default End Date for sale order '%s'.")
+                % self.display_name)
+
+    @api.onchange('default_start_date')
+    def default_start_date_change(self):
+        if (self.default_start_date and self.default_end_date and
+                self.default_start_date > self.default_end_date):
+            self.default_end_date = self.default_start_date
+
+    @api.onchange('default_end_date')
+    def default_end_date_change(self):
+        if (self.default_start_date and self.default_end_date and
+                self.default_start_date > self.default_end_date):
+            self.default_start_date = self.default_end_date
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    start_date = fields.Date(
+        string='Start Date', readonly=True,
+        states={'draft': [('readonly', False)]})
+    end_date = fields.Date(
+        string='End Date', readonly=True,
+        states={'draft': [('readonly', False)]})
+    number_of_days = fields.Integer(
+        compute='_compute_number_of_days',
+        inverse='_inverse_number_of_days',
+        string='Number of Days', readonly=False, store=True)
+    must_have_dates = fields.Boolean(related='product_id.must_have_dates')
+
+    @api.depends('start_date', 'end_date')
+    def _compute_number_of_days(self):
+        for line in self:
+            days = False
+            if line.start_date and line.end_date:
+                days = (line.end_date - line.start_date).days + 1
+            line.number_of_days = days
+
+    @api.onchange('number_of_days')
+    def _inverse_number_of_days(self):
+        res = {'warning': {}}
+        for line in self:
+            if line.number_of_days < 0:
+                res['warning']['title'] = _('Wrong number of days')
+                res['warning']['message'] = _(
+                    "On sale order line with product '%s', the "
+                    "number of days is negative (%d) ; this is not "
+                    "allowed. The number of days has been forced to 1.") % (
+                        line.product_id.display_name, line.number_of_days)
+                line.number_of_days = 1
+            if line.start_date:
+                line.end_date = line.start_date +\
+                    relativedelta(days=line.number_of_days - 1)
+            elif line.end_date:
+                line.start_date = line.end_date -\
+                    relativedelta(days=line.number_of_days - 1)
+        return res
+
+    @api.constrains('start_date', 'end_date')
```

Comment:
Not sure but as we are checking for `must_have_dates` here, IMO it should be added  here as well

Reply:
I disagree: we don't want to block setting "must_have_dates" for a product which was sold in the past.